File: modules/angular2/src/core/pipes/async_pipe.ts
import {isBlank, isPresent, isPromise, CONST, BaseException} from 'angular2/src/core/facade/lang';
import {Observable, Promise, ObservableWrapper} from 'angular2/src/core/facade/async';
import {Injectable} from 'angular2/di';

import {PipeTransform, PipeOnDestroy, WrappedValue} from 'angular2/change_detection';
import {InvalidPipeArgumentException} from './invalid_pipe_argument_exception';
import {ChangeDetectorRef} from 'angular2/change_detection';

import {Pipe} from '../metadata';


class ObservableStrategy {
  createSubscription(async: any, updateLatestValue: any): any {
    return ObservableWrapper.subscribe(async, updateLatestValue, e => { throw e; });
  }

  dispose(subscription: any): void { ObservableWrapper.dispose(subscription); }

  onDestroy(subscription: any): void { ObservableWrapper.dispose(subscription); }
}

class PromiseStrategy {
  createSubscription(async: any, updateLatestValue: any): any {
    return async.then(updateLatestValue);
  }

  dispose(subscription: any): void {}

  onDestroy(subscription: any): void {}
}

var _promiseStrategy = new PromiseStrategy();
var _observableStrategy = new ObservableStrategy();


/**
 * Implements async bindings to Observable and Promise.
 *
 * # Example
 *
 * In this example we bind the description observable to the DOM. The async pipe will convert an
 *observable to the
 * latest value it emitted. It will also request a change detection check when a new value is
 *emitted.
 *
 *  ```
 * @Component({
 *   selector: "task-cmp",
 *   changeDetection: ChangeDetectionStrategy.OnPush
 * })
 * @View({
 *   template: "Task Description {{ description | async }}"
 * })
 * class Task {
 *  description:Observable<string>;
 * }
 *
 * ```
 */
@Pipe({name: 'async'})
@Injectable()
export class AsyncPipe implements PipeTransform, PipeOnDestroy {
  _latestValue: Object = null;
  _latestReturnedValue: Object = null;

  _subscription: Object = null;
  _obj: Observable | Promise<any> = null;
  private _strategy: any = null;

  constructor(public _ref: ChangeDetectorRef) {}

  onDestroy(): void {
    if (isPresent(this._subscription)) {
      this._dispose();
    }
  }

  transform(obj: Observable | Promise<any>, args?: any[]): any {
    if (isBlank(this._obj)) {
      if (isPresent(obj)) {
        this._subscribe(obj);
      }
      return null;
    }

    if (obj !== this._obj) {
      this._dispose();
      return this.transform(obj);
    }

    if (this._latestValue === this._latestReturnedValue) {
      return this._latestReturnedValue;
    } else {
      this._latestReturnedValue = this._latestValue;
      return WrappedValue.wrap(this._latestValue);
    }
  }

  _subscribe(obj: Observable | Promise<any>): void {
    this._obj = obj;
    this._strategy = this._selectStrategy(obj);
    this._subscription =
        this._strategy.createSubscription(obj, value => this._updateLatestValue(obj, value));
  }

  _selectStrategy(obj: Observable | Promise<any>): any {
    if (isPromise(obj)) {
      return _promiseStrategy;
    } else if (ObservableWrapper.isObservable(obj)) {
      return _observableStrategy;
    } else {
      throw new InvalidPipeArgumentException(AsyncPipe, obj);
    }
  }

  _dispose(): void {
    this._strategy.dispose(this._subscription);
    this._latestValue = null;
    this._latestReturnedValue = null;
    this._subscription = null;
    this._obj = null;
  }

  _updateLatestValue(async: any, value: Object) {
    if (async === this._obj) {
      this._latestValue = value;
      this._ref.markForCheck();
    }
  }
}

File: lib/src/common/pipes/async_pipe.dart
library angular2.src.common.pipes.async_pipe;

import "package:angular2/src/facade/lang.dart"
    show isBlank, isPresent, isPromise;
import "package:angular2/src/facade/async.dart"
    show Future, ObservableWrapper, Stream, EventEmitter;
import "package:angular2/src/core/metadata.dart" show Pipe;
import "package:angular2/src/core/di.dart" show Injectable;
import "package:angular2/src/core/change_detection.dart"
    show ChangeDetectorRef, PipeOnDestroy, PipeTransform, WrappedValue;
import "invalid_pipe_argument_exception.dart" show InvalidPipeArgumentException;

class ObservableStrategy {
  dynamic createSubscription(dynamic async, dynamic updateLatestValue) {
    return ObservableWrapper.subscribe(async, updateLatestValue, (e) {
      throw e;
    });
  }

  void dispose(dynamic subscription) {
    ObservableWrapper.dispose(subscription);
  }

  void onDestroy(dynamic subscription) {
    ObservableWrapper.dispose(subscription);
  }
}

class PromiseStrategy {
  dynamic createSubscription(dynamic async, dynamic updateLatestValue) {
    return async.then(updateLatestValue);
  }

  void dispose(dynamic subscription) {}
  void onDestroy(dynamic subscription) {}
}

var _promiseStrategy = new PromiseStrategy();
var _observableStrategy = new ObservableStrategy();

/**
 * The `async` pipe subscribes to an Observable or Promise and returns the latest value it has
 * emitted.
 * When a new value is emitted, the `async` pipe marks the component to be checked for changes.
 *
 * ### Example
 * The example below binds the `time` Observable to the view. Every 500ms, the `time` Observable
 * updates the view with the current time.
 *
 * ```
 * import {Observable} from 'angular2/core';
 * @Component({
 *   selector: "task-cmp",
 *   template: "Time: {{ time | async }}"
 * })
 * class Task {
 *   time = new Observable<number>(observer => {
 *     setInterval(_ =>
 *       observer.next(new Date().getTime()), 500);
 *   });
 * }
 * ```
 */
@Pipe(name: "async", pure: false)
@Injectable()
class AsyncPipe implements PipeTransform, PipeOnDestroy {
  /** @internal */
  Object _latestValue = null;
  /** @internal */
  Object _latestReturnedValue = null;
  /** @internal */
  Object _subscription = null;
  /** @internal */
  dynamic /* Stream < dynamic > | Future < dynamic > | EventEmitter < dynamic > */ _obj =
      null;
  dynamic _strategy = null;
  /** @internal */
  ChangeDetectorRef _ref;
  AsyncPipe(ChangeDetectorRef _ref) {
    this._ref = _ref;
  }
  void onDestroy() {
    if (isPresent(this._subscription)) {
      this._dispose();
    }
  }

  dynamic transform(
      dynamic /* Stream < dynamic > | Future < dynamic > | EventEmitter < dynamic > */ obj,
      [List<dynamic> args]) {
    if (isBlank(this._obj)) {
      if (isPresent(obj)) {
        this._subscribe(obj);
      }
      return null;
    }
    if (!identical(obj, this._obj)) {
      this._dispose();
      return this.transform(obj);
    }
    if (identical(this._latestValue, this._latestReturnedValue)) {
      return this._latestReturnedValue;
    } else {
      this._latestReturnedValue = this._latestValue;
      return WrappedValue.wrap(this._latestValue);
    }
  }

  /** @internal */
  void _subscribe(
      dynamic /* Stream < dynamic > | Future < dynamic > | EventEmitter < dynamic > */ obj) {
    this._obj = obj;
    this._strategy = this._selectStrategy(obj);
    this._subscription = this._strategy.createSubscription(
        obj, (value) => this._updateLatestValue(obj, value));
  }

  /** @internal */
  dynamic _selectStrategy(
      dynamic /* Stream < dynamic > | Future < dynamic > | EventEmitter < dynamic > */ obj) {
    if (isPromise(obj)) {
      return _promiseStrategy;
    } else if (ObservableWrapper.isObservable(obj)) {
      return _observableStrategy;
    } else {
      throw new InvalidPipeArgumentException(AsyncPipe, obj);
    }
  }

  /** @internal */
  void _dispose() {
    this._strategy.dispose(this._subscription);
    this._latestValue = null;
    this._latestReturnedValue = null;
    this._subscription = null;
    this._obj = null;
  }

  /** @internal */
  _updateLatestValue(dynamic async, Object value) {
    if (identical(async, this._obj)) {
      this._latestValue = value;
      this._ref.markForCheck();
    }
  }
}
